app/src/main/java/com/github/wnder/picture/InternalCachePictureDatabase.java

```diff
@@ -0,0 +1,198 @@
+package com.github.wnder.picture;
+
+import android.content.Context;
+import android.graphics.Bitmap;
+import android.location.Location;
+import android.net.Uri;
+
+import com.github.wnder.NetworkInformation;
+import com.github.wnder.WnderApplication;
+
+import org.json.JSONException;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import javax.inject.Inject;
+
+/**
+ * Class that links firebase db and local db depending on internet connection availability
+ */
+public class InternalCachePictureDatabase implements PicturesDatabase{
+    private final FirebasePicturesDatabase remoteDatabase;
+    private final LocalPictureDatabase localDatabase;
+    private Context context;
+
+    /**
+     * Constructor
+     * @param context app context
+     */
+    @Inject
+    public InternalCachePictureDatabase(Context context){
+        remoteDatabase = new FirebasePicturesDatabase();
+        localDatabase = new LocalPictureDatabase(context);
+        this.context = context;
+    }
+
+    /**
+     * Checks if app is online or not
+     * @return true if available internet connection, false o/w
+     */
+    public boolean isOnline(){
+        return NetworkInformation.isNetworkAvailable(context);
+    }
+
+    @Override
+    public CompletableFuture<Location> getLocation(String uniqueId) {
+        //If connection available, remote db
+        if (isOnline()) {
+            return remoteDatabase.getLocation(uniqueId);
+        }
+        //else, local db
+        else {
+            CompletableFuture<Location> cf = new CompletableFuture<>();
+            cf.complete(localDatabase.getLocation(uniqueId));
+            return cf;
+        }
+    }
+
+    @Override
+    public CompletableFuture<Location> getApproximateLocation(String uniqueId) throws IllegalStateException{
+        //If connection available, remote db
+        if (isOnline()) {
+            return remoteDatabase.getApproximateLocation(uniqueId);
+        }
+        //Not available when no internet
+        else {
+            throw new IllegalStateException("The approximate location is not available locally");
```

Comment:
I think we could however it doesn't make much sense as you said because not only is the computation done only in the remote db class but also we only does this computation when we are guessing a picture (and not anymore once it's guessed)